content/courses/advanced-ada/chapters/types.rst

```diff
@@ -3201,19 +3201,448 @@ majority of the cases, you don't need it at all |mdash| except for special
 cases such as when interfacing with C code that makes use of union types or
 solving very specific problems when doing low-level programming.
 
-Variable control
-----------------
+Shared variable control
+-----------------------
 
-.. admonition:: Relevant topics
+Ada has built-in support for handling both volatile and atomic data. Let's
+start by discussing volatile objects.
 
-    - **Briefly** discuss :ada:`Atomic`, :ada:`Volatile`, etc
-    - `Shared Variable Control <http://www.ada-auth.org/standards/2xrm/html/RM-C-6.html>`_
-    - `The Package System.Atomic_Operations <http://www.ada-auth.org/standards/2xrm/html/RM-C-6-1.html>`_
+Volatile
+~~~~~~~~
 
-.. todo::
+A `volatile <https://en.wikipedia.org/wiki/Volatile_(computer_programming)>`_
+object can be described as an object in memory whose value may change between
+two consecutive memory accesses of a process A |mdash| even if process A itself
+hasn't changed the value. This situation may arise when an object in memory is
+being shared by multiple threads. For example, a thread *B* may modify the
+value of that object between two read accesses of a thread *A*. Another typical
+example is the one of
+`memory-mapped I/O <https://en.wikipedia.org/wiki/Memory-mapped_I/O>`_, where
+the hardware might be constantly changing the value of an object in memory.
 
-    Complete section!
+Because the value of a volatile object may be constantly changing, a compiler
+cannot generate code that stores the value of that object into a register and
```

Comment:
"cannot generate code that stores the value of that object into a register and" =>
"cannot generate code to store the value of that object in a register and then"

(This is a tricky bit of grammar.)

Reply:
Thanks a lot!